misc/discoverynode/src/udmi/discovery/ether.py

```diff
@@ -0,0 +1,173 @@
+    getattr(self, f"{self.last_bathometer_reading}_stop_discovery")()
+
+  def magic_bathometer(self, depth: str) -> str:
+    """Identifies the discovery scan to run based on the given depth"""
+    match depth:
+      case "ping":
+        return "ping"
+      case "ports":
+        return "nmap"
+      case _:
+        raise RuntimeError(f"unmatched depth {depth}")
+
+  @discovery.catch_exceptions_to_state
+  @discovery.main_task
+  def ping_dispatcher(self, target_ips: list[str]):
+    """Dispatches ping tasks across simulateous workers"""
+    if not target_ips:
+      logging.warning("no targets given for ping scan")
+      return
+    with concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
+      futures = [executor.submit(self.ping_task, ip) for ip in target_ips]
+      while (
+          concurrent.futures.wait(futures, 1)
+          != concurrent.futures.ALL_COMPLETED
+      ):
+
+        if self.cancel_threads.is_set():
+          executor.shutdown(True, cancel_futures=True)
+          # cancelled futures return as not completed, so the loop must exit
+          break
+
+  def ping_task(self, target_ip: str) -> bool:
+    """Sends a single ping to the target IP. Returns True if successful."""
+    logging.debug("ping task %s", target_ip)
+    try:
+      # -c 1: expect 1 packet
+      # -W 2: wait 2 seconds for a response
+      result = subprocess.run(
+          ["/usr/bin/ping", "-c", "1", "-2", "2", target_ip],
+          stdout=subprocess.PIPE,
+          stderr=subprocess.STDOUT,
+          encoding="utf-8",
+          check=True,
+          timeout=5,
+      )
+      event = udmi.schema.discovery_event.DiscoveryEvent(
+          generation=self.generation,
+          family=self.family,
+          addr=target_ip,
+      )
+      self.publish(event)
+
+    except subprocess.CalledProcessError as e:
+      logging.error(f"Ping failed for %s: %s", target_ip, e.output)
+      return False
+    except Exception as e:
+      logging.error(f"Ping exception for %s: %s", target_ip, e)
+      return False
+
+  def ping_start_discovery(self):
+    """Start ping discovery scan."""
+    self.cancel_threads.clear()
+    self.ping_thread = threading.Thread(
+        target=self.ping_dispatcher, args=[self.config.addrs], daemon=True
+    )
+    self.ping_thread.start()
+
+  def ping_stop_discovery(self):
+    """Stop ping discovery scan."""
+    self.cancel_threads.set()
+    if self.ping_thread and self.ping_thread.is_alive():
+      self.ping_thread.join()
+
+  def nmap_start_discovery(self):
+    self.cancel_threads.clear()
+    self.nmap_thread = threading.Thread(
+        target=self.nmap_runner, args=[], daemon=True
+    )
+    self.nmap_thread.start()
+
+  def nmap_stop_discovery(self):
+    logging.info("stopping")
+    self.cancel_threads.set()
+    self.nmap_thread.join()
+
+  @discovery.catch_exceptions_to_state
+  @discovery.main_task
+  def nmap_runner(self):
```

Comment:
note to self: change this to take config.addrs
note to reviewrs: this all used to be in `nmap.py` which was renamed